### ISSUE TITLE
using *static.lib to connect MQTTAsync but failed and return error code 6

### ISSUE THREAD

**Issue by jaysontatum87** [eclipse-paho/paho.mqtt.c#925]
**Problem**
I tried to use paho-mqtt3as-static.lib to connect mqtt async ssl but failed and return error code 6, I can't figure out the reason even  I read the source code in "MQTTAsync.c".
The paho-mqtt3as-static.lib is complied by using [paho.mqtt Version 1.3.0 (Paho 1.4)](https://github.com/eclipse/paho.mqtt.c/releases/tag/v1.3.0)
I succeed to connect "tcp//test.mosquitto.org:1883 "[MQTT server/broker](https://test.mosquitto.org/) by the test project MQTTAsync_subscribe with the complied paho-mqtt3a.lib. 
And also I can connect "ssl://test.mosquitto.org:8883" as follow code by using paho-mqtt3as.lib. 
so I think  it also can connect "ssl://test.mosquitto.org:8883" by using paho-mqtt3as-static.lib, but it failed.
I also tried to connect mqtt async without ssl by using "paho-mqtt3a-static.lib", it also return error code 6. It seems like the problem caused by "*-static.lib".
Is it "*-static.lib" cause the problem? and how can I solve it. 
Thanks for your attention!

**Here are my test code.**

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "MQTTAsync.h"

#if !defined(WIN32)
#include <unistd.h>
#else
#include <windows.h>
#endif

#if defined(_WRS_KERNEL)
#include <OsWrapper.h>
#endif

#define ADDRESS    "ssl://test.mosquitto.org:8883"
#define CLIENTID    "ExampleClientPub"
#define TOPIC       "MQTT Examples"
#define PAYLOAD     "Hello World!"
#define QOS         1
#define TIMEOUT     10000L

volatile MQTTAsync_token deliveredtoken;

int finished = 0;

int main(int argc, char* argv[])
{

	MQTTAsync client;
	MQTTAsync_connectOptions conn_opts = MQTTAsync_connectOptions_initializer;
	int rc;

	MQTTAsync_create(&client, ADDRESS, CLIENTID, MQTTCLIENT_PERSISTENCE_NONE, NULL);

	conn_opts.keepAliveInterval = 20;
	conn_opts.cleansession = 1;
	conn_opts.context = client;

	MQTTAsync_SSLOptions ssl_opts = MQTTAsync_SSLOptions_initializer;
	conn_opts.ssl = (MQTTAsync_SSLOptions *)malloc(sizeof(MQTTAsync_SSLOptions));
	conn_opts.ssl = &ssl_opts;
	conn_opts.ssl->verify = 0;
	conn_opts.ssl->trustStore = "C:/Users/user/Desktop/mqtt/src/samples/Debug/mosquitto.org.crt";
	if ((rc = MQTTAsync_connect(client, &conn_opts)) != MQTTASYNC_SUCCESS)
	{
		printf("Failed to start connect, return code %d\n", rc);
		exit(EXIT_FAILURE);
	}

	printf("Waiting for publication of %s\n"
         "on topic %s for client with ClientID: %s\n",
         PAYLOAD, TOPIC, CLIENTID);
	while (!finished)
		#if defined(WIN32)
			Sleep(100);
		#else
			usleep(10000L);
		#endif

	MQTTAsync_destroy(&client);
 	return rc;
}

**Comment by jaysontatum87** [comment]
check the return value of API MQTTAsync_connect and get as follow:
<b>6-255</b>: Reserved for future use<br>

**Comment by fpagliughi** [comment]
The error is: "A NULL parameter has been supplied when this is invalid."

See `MQTTAsync.h`:  
https://github.com/eclipse/paho.mqtt.c/blob/8d3d8276a77f27b360000bc3a21b47f93f810d0f/src/MQTTAsync.h#L137-L140

It looks like you're setting up your option structs a little strangely. You declare one SSL options struct on the stack then malloc a second one on the heap. In the end you're using the stack one, which seems good, but perhaps some additional options are required.

**Comment by fpagliughi** [comment]
And I was assuming that the error code was actually -6, not +6.

**Comment by jaysontatum87** [comment]
Thank you for your reply. 
The error code was actually +6. 
I commented out malloc link like this `//conn_opts.ssl = (MQTTAsync_SSLOptions *)malloc(sizeof(MQTTAsync_SSLOptions));`
and also returned  the error code +6.

I complied the mqtt 1.3.4 paho-mqtt3as-static.lib  rather than now 1.3.0 and connected successfully.So I think this problem may be caused by mqtt version.
And I also found out memoryleak as follow by using _CrtDumpMemoryLeaks().

**Detected memory leaks!
Dumping objects ->
{8137} normal block at 0x00C1C3C8, 12 bytes long.
 Data: <            > 98 C1 C1 00 00 00 00 00 0D 00 00 00 
{8136} normal block at 0x00C238F0, 16 bytes long.
 Data: <                > 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
{8135} normal block at 0x00C235F0, 20 bytes long.
 Data: <     8          > 00 00 00 00 F0 38 C2 00 00 00 00 00 04 00 00 00 
{8134} normal block at 0x00C1C198, 12 bytes long.
 Data: <     5      > 0D 00 00 00 F0 35 C2 00 00 00 00 00 
{7818} normal block at 0x00C1C358, 12 bytes long.
 Data: <            > F8 C5 C1 00 00 00 00 00 06 00 00 00 
{7817} normal block at 0x00C23A70, 16 bytes long.
 Data: <                > 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
{7816} normal block at 0x00C23B70, 20 bytes long.
 Data: <    p:          > 00 00 00 00 70 3A C2 00 00 00 00 00 04 00 00 00 
{7815} normal block at 0x00C1C5F8, 12 bytes long.
 Data: <    p;      > 06 00 00 00 70 3B C2 00 00 00 00 00 
{7248} normal block at 0x00BE5578, 12 bytes long.
 Data: <([          > 28 5B BE 00 00 00 00 00 03 00 00 00 
{7247} normal block at 0x00C05638, 16 bytes long.
 Data: <                > 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
{7246} normal block at 0x00C054F8, 20 bytes long.
 Data: <    8V          > 00 00 00 00 38 56 C0 00 00 00 00 00 04 00 00 00 
{7245} normal block at 0x00BE5B28, 12 bytes long.
 Data: <     T      > 03 00 00 00 F8 54 C0 00 00 00 00 00 
{7223} normal block at 0x00BE5AF0, 12 bytes long.
 Data: <p           > 70 81 C0 00 00 00 00 00 0C DF 03 00 
{7222} normal block at 0x00BFEF68, 64 bytes long.
 Data: <                > 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
{7221} normal block at 0x00C02408, 96 bytes long.
 Data: <h   p           > 68 EF BF 00 70 B1 1F 00 80 B1 1F 00 08 00 00 00 
{7220} normal block at 0x00C08170, 400 bytes long.
 Data: <    <L          > 00 00 00 00 3C 4C 00 00 00 00 00 00 00 00 00 00 
{6958} normal block at 0x00BE5A48, 12 bytes long.
 Data: < U          > E8 55 BE 00 00 00 00 00 0A 00 00 00 
{6957} normal block at 0x00BBF5F0, 16 bytes long.
 Data: <                > 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
{6956} normal block at 0x00BEF650, 20 bytes long.
 Data: <                > 00 00 00 00 F0 F5 BB 00 00 00 00 00 04 00 00 00 
{6955} normal block at 0x00BE55E8, 12 bytes long.
 Data: <    P       > 0A 00 00 00 50 F6 BE 00 00 00 00 00 
{6920} normal block at 0x00BECFC0, 12 bytes long.
 Data: <p           > 70 CE BE 00 00 00 00 00 00 00 00 00 
{6919} normal block at 0x00BEFF90, 16 bytes long.
 Data: <                > 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
{6918} normal block at 0x00BF0390, 20 bytes long.
 Data: <                > 00 00 00 00 90 FF BE 00 00 00 00 00 04 00 00 00 
{6917} normal block at 0x00BECE70, 12 bytes long.
 Data: <            > 00 00 00 00 90 03 BF 00 00 00 00 00 
{3562} normal block at 0x00BED180, 12 bytes long.
 Data: <    HZ      > 00 CE BE 00 48 5A BE 00 02 00 00 00 
{3561} normal block at 0x00BEFD50, 16 bytes long.
 Data: <                > 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
{3560} normal block at 0x00BF0190, 20 bytes long.
 Data: <    P           > 00 00 00 00 50 FD BE 00 00 00 00 00 04 00 00 00 
{3559} normal block at 0x00BECE00, 12 bytes long.
 Data: <            > 02 00 00 00 90 01 BF 00 00 00 00 00 
{3541} normal block at 0x00BED340, 12 bytes long.
 Data: <            > C8 CD BE 00 00 00 00 00 04 00 00 00 
{3540} normal block at 0x00BEFE10, 16 bytes long.
 Data: <                > 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
{3539} normal block at 0x00BEFDD0, 20 bytes long.
 Data: <                > 00 00 00 00 10 FE BE 00 00 00 00 00 04 00 00 00 
{3538} normal block at 0x00BECDC8, 12 bytes long.
 Data: <            > 04 00 00 00 D0 FD BE 00 00 00 00 00 
{3526} normal block at 0x00BEFF10, 20 bytes long.
 Data: <    x66         > 00 00 00 00 78 36 36 00 00 00 00 00 00 00 00 00 
{3525} normal block at 0x00BED110, 12 bytes long.
 Data: <0           > 30 D0 BE 00 C8 C3 C1 00 05 00 00 00 
{3524} normal block at 0x00BF0410, 16 bytes long.
 Data: <                > 10 FF BE 00 00 00 00 00 00 00 00 00 00 00 00 00 
{3523} normal block at 0x00BF0350, 20 bytes long.
 Data: <                > 01 00 00 00 10 04 BF 00 00 00 00 00 04 00 00 00 
{3522} normal block at 0x00BED030, 12 bytes long.
 Data: <    P       > 05 00 00 00 50 03 BF 00 01 00 00 00 
{3435} normal block at 0x00BEFAD0, 20 bytes long.
 Data: <     G7         > 00 00 00 00 A8 47 37 00 00 00 00 00 00 00 00 00 
{3434} normal block at 0x00BED308, 12 bytes long.
 Data: <            > F0 D1 BE 00 00 00 00 00 01 00 00 00 
{3433} normal block at 0x00BEF850, 16 bytes long.
 Data: <                > D0 FA BE 00 00 00 00 00 00 00 00 00 00 00 00 00 
{3432} normal block at 0x00BEF810, 20 bytes long.
 Data: <    P           > 01 00 00 00 50 F8 BE 00 00 00 00 00 04 00 00 00 
{3431} normal block at 0x00BED1F0, 12 bytes long.
 Data: <            > 01 00 00 00 10 F8 BE 00 01 00 00 00 
{3430} normal block at 0x00BF0ED8, 64 bytes long.
 Data: <            xU  > C0 CF BE 00 08 D3 BE 00 80 D1 BE 00 78 55 BE 00 
{3429} normal block at 0x00BC0650, 96 bytes long.
 Data: <                > D8 0E BF 00 90 86 20 00 A0 86 20 00 08 00 00 00 
{243} normal block at 0x00BBF6B0, 16 bytes long.
 Data: <                > 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
{242} normal block at 0x00BBF8B0, 20 bytes long.
 Data: <                > 00 00 00 00 B0 F6 BB 00 01 00 00 00 04 00 00 00 
Object dump complete.**

if I connect mqtt async without ssl that no memoryleak.
Is it a bug of paho-mqtt3as-static.lib?

**Comment by fpagliughi** [comment]
How are you getting out of the application? I see nothing that sets the "finished" flag.

**Comment by jaysontatum87** [comment]
1.
After subscribe succeeded, click "Q" to the "finished" flag and  get out from the application.
I was sorry to just post the part of test code,
The entire test code is as follow:

```
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "MQTTAsync.h"

#if !defined(WIN32)
#include <unistd.h>
#else
#include <windows.h>
#endif

#if defined(_WRS_KERNEL)
#include <OsWrapper.h>
#endif

#define ADDRESS    "ssl://test.mosquitto.org:8883"
#define CLIENTID    "ExampleClientSub"
#define TOPIC       "MQTT Examples"
#define PAYLOAD     "Hello World!"
#define QOS         1
#define TIMEOUT     10000L


#include <crtdbg.h>
#ifdef _DEBUG
#define   new                   new(_NORMAL_BLOCK, __FILE__, __LINE__)
#define   malloc(s)             _malloc_dbg(s, _NORMAL_BLOCK, __FILE__, __LINE__)
#define   calloc(c, s)          _calloc_dbg(c, s, _NORMAL_BLOCK, __FILE__, __LINE__)
#define   realloc(p, s)         _realloc_dbg(p, s, _NORMAL_BLOCK, __FILE__, __LINE__)
#define   _recalloc(p, c, s)    _recalloc_dbg(p, c, s, _NORMAL_BLOCK, __FILE__, __LINE__)
#define   _expand(p, s)         _expand_dbg(p, s, _NORMAL_BLOCK, __FILE__, __LINE__)
#endif

volatile MQTTAsync_token deliveredtoken;

int disc_finished = 0;
int subscribed = 0;
int finished = 0;

void connlost(void *context, char *cause)
{
	MQTTAsync client = (MQTTAsync)context;
	MQTTAsync_connectOptions conn_opts = MQTTAsync_connectOptions_initializer;
	int rc;

	printf("\nConnection lost\n");
	if (cause)
		printf("     cause: %s\n", cause);

	printf("Reconnecting\n");
	conn_opts.keepAliveInterval = 20;
	conn_opts.cleansession = 1;
	if ((rc = MQTTAsync_connect(client, &conn_opts)) != MQTTASYNC_SUCCESS)
	{
		printf("Failed to start connect, return code %d\n", rc);
		finished = 1;
	}
}


int msgarrvd(void *context, char *topicName, int topicLen, MQTTAsync_message *message)
{
	int i;
	char* payloadptr;

	printf("Message arrived\n");
	printf("     topic: %s\n", topicName);
	printf("   message: ");

	payloadptr = message->payload;
	for (i = 0; i < message->payloadlen; i++)
	{
		putchar(*payloadptr++);
	}
	putchar('\n');
	MQTTAsync_freeMessage(&message);
	MQTTAsync_free(topicName);
	return 1;
}


void onDisconnect(void* context, MQTTAsync_successData* response)
{
	printf("Successful disconnection\n");
	disc_finished = 1;
}


void onSubscribe(void* context, MQTTAsync_successData* response)
{
	printf("Subscribe succeeded\n");
	subscribed = 1;
}

void onSubscribeFailure(void* context, MQTTAsync_failureData* response)
{
	printf("Subscribe failed, rc %d\n", response ? response->code : 0);
	finished = 1;
}


void onConnectFailure(void* context, MQTTAsync_failureData* response)
{
	printf("Connect failed, rc %d\n", response ? response->code : 0);
	finished = 1;
}


void onConnect(void* context, MQTTAsync_successData* response)
{
	MQTTAsync client = (MQTTAsync)context;
	MQTTAsync_responseOptions opts = MQTTAsync_responseOptions_initializer;
	int rc;

	printf("Successful connection\n");

	printf("Subscribing to topic %s\nfor client %s using QoS%d\n\n"
		"Press Q<Enter> to quit\n\n", TOPIC, CLIENTID, QOS);
	opts.onSuccess = onSubscribe;
	opts.onFailure = onSubscribeFailure;
	opts.context = client;

	deliveredtoken = 0;

	if ((rc = MQTTAsync_subscribe(client, TOPIC, QOS, &opts)) != MQTTASYNC_SUCCESS)
	{
		printf("Failed to start subscribe, return code %d\n", rc);
		exit(EXIT_FAILURE);
	}
}


int main(int argc, char* argv[])
{
	//MQTTAsync_init_options init_opts = MQTTAsync_init_options_initializer;
	//MQTTAsync_global_init(&init_opts);
	MQTTAsync client;
	MQTTAsync_connectOptions conn_opts = MQTTAsync_connectOptions_initializer;
	MQTTAsync_disconnectOptions disc_opts = MQTTAsync_disconnectOptions_initializer;
	int rc;
	int ch;

	MQTTAsync_create(&client, ADDRESS, CLIENTID, MQTTCLIENT_PERSISTENCE_NONE, NULL);

	MQTTAsync_setCallbacks(client, client, connlost, msgarrvd, NULL);

	conn_opts.keepAliveInterval = 20;
	conn_opts.cleansession = 1;
	conn_opts.onSuccess = onConnect;
	conn_opts.onFailure = onConnectFailure;
	conn_opts.context = client;
	MQTTAsync_SSLOptions ssl_opts = MQTTAsync_SSLOptions_initializer;
	//conn_opts.ssl = (MQTTAsync_SSLOptions *)malloc(sizeof(MQTTAsync_SSLOptions));
	conn_opts.ssl = &ssl_opts;
	conn_opts.ssl->verify = 0;
	conn_opts.ssl->trustStore = "mosquitto.org.crt";
	if ((rc = MQTTAsync_connect(client, &conn_opts)) != MQTTASYNC_SUCCESS)
	{
		printf("Failed to start connect, return code %d\n", rc);
		exit(EXIT_FAILURE);
	}

	while (!subscribed)
#if defined(WIN32)
		Sleep(100);
#else
		usleep(10000L);
#endif

	if (finished)
		goto exit;

	do
	{
		ch = getchar();
	} while (ch != 'Q' && ch != 'q');

	disc_opts.onSuccess = onDisconnect;
	if ((rc = MQTTAsync_disconnect(client, &disc_opts)) != MQTTASYNC_SUCCESS)
	{
		printf("Failed to start disconnect, return code %d\n", rc);
		exit(EXIT_FAILURE);
	}
	while (!disc_finished)
#if defined(WIN32)
		Sleep(100);
#else
		usleep(10000L);
#endif

exit:
	MQTTAsync_destroy(&client);
	_CrtDumpMemoryLeaks();
	return rc;
}

```

2.console result
I get the connected result from console like this:  
Successful connection
Subscribing to topic MQTT Examples
for client ExampleClientSub using QoS1

Press Q<Enter> to quit

Subscribe succeeded
Q
Successful disconnection


3.To fix memory leaks problem
by the way,I tried to use MQTTAsync_global_init(&init_opts) to solve the Detectedmemory leaks but got error like this;
```
SSL_CTX ctx is a nullptr when run SSL_CTX_set_info_callback in [openssl-1.0.2u\ssl\ssl_sess.c](https://github.com/openssl/openssl/releases/tag/OpenSSL_1_0_2u)
void SSL_CTX_set_info_callback(SSL_CTX *ctx,
                               void (*cb) (const SSL *ssl, int type, int val))
{
    ctx->info_callback = cb;
}
```

**Comment by icraggs** [comment]
There is already memory tracking in the library if you turn the tracing on as described in the readme.  If you run that and it reports memory unfreed it will tell you where from.

Also, there are some OpenSSL internal memory blocks that will not be freed.  This is because some applications use OpenSSL themselves so closing OpenSSL completely disrupts those applications.  I removed the functions to close OpenSSL completely in response to an issue.

**Comment by jaysontatum87** [comment]
Thanks for your reminder.

I used "maximum" trace_level and  didn't find any information like "memory leak" or "Some memory not freed at shutdown".
So  can I understand that this is not a memory leak  in paho-mqtt3as.lib and "unfreed memory" is caused by the part of unfreed OpenSSL internal memory blocks?

Is it necessary to free them manually? or keep it that way.

**Comment by jumoog** [comment]
you can try valgrind to see the memory leaks

**Comment by jaysontatum87** [comment]
> you can try valgrind to see the memory leaks

I test mqtt in win32 with "paho-mqtt3as.lib", so maybe it can not dectect  memory leaks by  valgrind.

There is not enough examples to refer to this OpenSSL unfreed memory problem.
I want to figure out  is it necessary to free  "unfreed memory"  caused by  OpenSSL internal memory blocks
manually after MQTTAsync_destroy(&client)?

**Comment by icraggs** [comment]
You can try using the functions in this list:

https://wiki.openssl.org/index.php/Library_Initialization#Cleanup

if this is concerning you and let us know what works and could be added to a general library terminate function, if one is added.

When the process is closed all memory associated with it will be cleaned up.

**Comment by cheryF** [comment]
If you use paho as static lib, manual call the "MQTTAsync_global_init" function to initialization mutex handlers.

MQTTAsync_lock_mutex(NULL); -> error code 6

**Comment by icraggs** [comment]
I think this is done.  There are other issues about global init/erminate.

**Comment by jaysontatum87** [comment]
Very sorry for the late reply.
I had tried the methods you mentioned on https://wiki.openssl.org/index.php/Library_Initialization#Cleanup
but it was not fixed.
It is fixed with #316 on 1.3.9?
